### PR TITLE
Handle FileNotFound for symlinks when using polling

### LINKED ISSUE #56190: Microsoft.Extensions.FileProviders.Physical.Tests disabled on Unix systems
https://github.com/dotnet/runtime/pull/56183/checks?check_run_id=3138813912
https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-56183-merge-f78074ea8e6047b2a6/Microsoft.Extensions.FileProviders.Physical.Tests/console.ac3f3b03.log?sv=2019-07-07&se=2021-08-11T22%3A22%3A03Z&sr=c&sp=rl&sig=ywBXxPkq8EuFYEZl2jIf5dvaPIqJDBOgm1Phou2QmU0%3D

```
Console log: 'Microsoft.Extensions.FileProviders.Physical.Tests' from job f78074ea-8e60-47b2-a6e4-961f30f8afa1 workitem 7254ea10-54b1-48a3-9672-5fbd419dfa26 (osx.1013.amd64.open) executed on machine dci-mac-build-083.local
+ ./RunTests.sh --runtime-path /tmp/helix/working/AB72097D/p
----- start Thu Jul 22 15:24:05 PDT 2021 =============== To repro directly: =====================================================
pushd .
/tmp/helix/working/AB72097D/p/dotnet exec --runtimeconfig Microsoft.Extensions.FileProviders.Physical.Tests.runtimeconfig.json --depsfile Microsoft.Extensions.FileProviders.Physical.Tests.deps.json xunit.console.dll Microsoft.Extensions.FileProviders.Physical.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing 
popd
===========================================================================================================
/private/tmp/helix/working/AB72097D/w/A0CB0914/e /private/tmp/helix/working/AB72097D/w/A0CB0914/e
  Discovering: Microsoft.Extensions.FileProviders.Physical.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  Microsoft.Extensions.FileProviders.Physical.Tests (found 76 of 92 test cases)
  Starting:    Microsoft.Extensions.FileProviders.Physical.Tests (parallel test collections = on, max threads = 4)
Unhandled exception. System.IO.FileNotFoundException: Could not find file '/var/folders/vn/knn7pvpd2y12gn_qhnnzqps80000gy/T/nnep5noo.kpg/e4a4s2kc.z55'.
File name: '/var/folders/vn/knn7pvpd2y12gn_qhnnzqps80000gy/T/nnep5noo.kpg/e4a4s2kc.z55'
   at System.IO.FileSystem.ResolveLinkTarget(String linkPath, Boolean returnFinalTarget, Boolean isDirectory) in /_/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs:line 603
   at Microsoft.Extensions.FileProviders.Physical.FileSystemInfoHelper.GetFileLinkTargetLastWriteTimeUtc(FileInfo fileInfo) in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/Internal/FileSystemInfoHelper.cs:line 54
   at Microsoft.Extensions.FileProviders.Physical.PollingFileChangeToken.GetLastWriteTimeUtc() in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PollingFileChangeToken.cs:line 57
   at Microsoft.Extensions.FileProviders.Physical.PollingFileChangeToken.get_HasChanged() in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PollingFileChangeToken.cs:line 101
   at Microsoft.Extensions.FileProviders.Physical.PhysicalFilesWatcher.RaiseChangeEvents(Object state) in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PhysicalFilesWatcher.cs:line 437
   at System.Threading.TimerQueueTimer.Fire(Boolean isThreadPool) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs:line 673
   at System.Threading.TimerQueue.FireNextTimers() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs:line 326
```

### LINKED ISSUE #56643: [libraries] CoreCLR Windows x86 Microsoft.Extensions.FileProviders.Physical.Tests FileNotFoundException
```
Unhandled exception. System.IO.FileNotFoundException: Could not find file 'C:\h\w\A55B08F1\t\nf1hpxkl.qvb\uw23btx3.aw4'.
File name: 'C:\h\w\A55B08F1\t\nf1hpxkl.qvb\uw23btx3.aw4'
   at System.IO.FileSystem.GetFindData(String fullPath, Boolean isDirectory, WIN32_FIND_DATA& findData) in /_/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs:line 214
   at System.IO.FileSystem.GetFinalLinkTarget(String linkPath, Boolean isDirectory) in /_/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs:line 533
   at System.IO.FileSystem.ResolveLinkTarget(String linkPath, Boolean returnFinalTarget, Boolean isDirectory) in /_/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs:line 428
   at System.IO.FileSystemInfo.ResolveLinkTarget(Boolean returnFinalTarget) in /_/src/libraries/System.Private.CoreLib/src/System/IO/FileSystemInfo.cs:line 170
   at Microsoft.Extensions.FileProviders.Physical.FileSystemInfoHelper.GetFileLinkTargetLastWriteTimeUtc(FileInfo fileInfo) in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/Internal/FileSystemInfoHelper.cs:line 54
   at Microsoft.Extensions.FileProviders.Physical.PollingFileChangeToken.GetLastWriteTimeUtc() in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PollingFileChangeToken.cs:line 57
   at Microsoft.Extensions.FileProviders.Physical.PollingFileChangeToken.get_HasChanged() in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PollingFileChangeToken.cs:line 101
   at Microsoft.Extensions.FileProviders.Physical.PhysicalFilesWatcher.RaiseChangeEvents(Object state) in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PhysicalFilesWatcher.cs:line 455
   at System.Threading.TimerQueueTimer.CallCallback(Boolean isThreadPool) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs:line 693
   at System.Threading.TimerQueueTimer.Fire(Boolean isThreadPool) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs:line 672
   at System.Threading.TimerQueue.FireNextTimers() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs:line 327
   at System.Threading.TimerQueue.AppDomainTimerCallback(Int32 id) in /_/src/coreclr/System.Private.CoreLib/src/System/Threading/Timer.CoreCLR.cs:line 59
```

Build: https://dev.azure.com/dnceng/public/_build/results?buildId=1268612&view=logs&j=457f7e88-dfa2-5bd9-f871-fdf124c2477d&t=bfe52dfb-2099-5c7f-ee52-70a1d81c544e
Log: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-56583-merge-6bcb853222e6486085/Microsoft.Extensions.FileProviders.Physical.Tests/console.59a2c30d.log?sv=2019-07-07&se=2021-08-19T16%3A20%3A08Z&sr=c&sp=rl&sig=fxnMmJrGTiPmf8Cc4vJ7GXCiefdJwx34Js7FXZA1ZqY%3D

### LINKED ISSUE #57128: Microsoft.Extensions.FileProviders.Physical.Tests terminated
Product bug here. FileNotFoundException should not take down the app.

'Microsoft.Extensions.FileProviders.Physical.Tests' from job 7359fb7e-bed2-4677-8025-0794e6d50356 workitem 38c6c2ba-c9ab-42a5-998b-a79c5089ce69 (windows.10.amd64.server19h1.open.rt) executed on machine a0036MX

net6.0-windows-Debug-x86-CoreCLR_release-Windows.10.Amd64.Server19H1.Open

```

Starting:    Microsoft.Extensions.FileProviders.Physical.Tests (parallel test collections = on, max threads = 2)
Unhandled exception. System.IO.FileNotFoundException: Could not find file 'C:\h\w\A9AD08CB\t\ev1xwosb.xvy\igyuv4fj.oko'.
File name: 'C:\h\w\A9AD08CB\t\ev1xwosb.xvy\igyuv4fj.oko'
   at System.IO.FileSystem.GetFindData(String fullPath, Boolean isDirectory, Boolean ignoreAccessDenied, WIN32_FIND_DATA& findData) in /_/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs:line 219
   at System.IO.FileSystem.GetFinalLinkTarget(String linkPath, Boolean isDirectory) in /_/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs:line 538
   at System.IO.FileSystem.ResolveLinkTarget(String linkPath, Boolean returnFinalTarget, Boolean isDirectory) in /_/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs:line 433
   at System.IO.FileSystemInfo.ResolveLinkTarget(Boolean returnFinalTarget) in /_/src/libraries/System.Private.CoreLib/src/System/IO/FileSystemInfo.cs:line 170
   at Microsoft.Extensions.FileProviders.Physical.FileSystemInfoHelper.GetFileLinkTargetLastWriteTimeUtc(FileInfo fileInfo) in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/Internal/FileSystemInfoHelper.cs:line 54
   at Microsoft.Extensions.FileProviders.Physical.PollingFileChangeToken.GetLastWriteTimeUtc() in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PollingFileChangeToken.cs:line 57
   at Microsoft.Extensions.FileProviders.Physical.PollingFileChangeToken.get_HasChanged() in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PollingFileChangeToken.cs:line 101
   at Microsoft.Extensions.FileProviders.Physical.PhysicalFilesWatcher.RaiseChangeEvents(Object state) in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PhysicalFilesWatcher.cs:line 455
   at System.Threading.TimerQueueTimer.CallCallback(Boolean isThreadPool) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs:line 712
   at System.Threading.TimerQueueTimer.Fire(Boolean isThreadPool) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs:line 673
   at System.Threading.TimerQueue.FireNextTimers() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs:line 326
   at System.Threading.TimerQueue.AppDomainTimerCallback(Int32 id) in /_/src/coreclr/System.Private.CoreLib/src/System/Threading/Timer.CoreCLR.cs:line 59
----- end Tue 08/10/2021  5:45:17.51 ----- exit code -532462766 ------------
```

### PR DESCRIPTION
**Pull request by jozkee**
Fixes https://github.com/dotnet/runtime/issues/56190
Fixes https://github.com/dotnet/runtime/issues/56643
Fixes #57128

### CODE REVIEW

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you @Jozkee

**Review by jeffhandley — COMMENTED** [review]
👍

File: src/libraries/Microsoft.Extensions.FileProviders.Physical/src/Internal/FileSystemInfoHelper.cs [hunk 1]
```diff
@@ -45,16 +45,28 @@ public static bool IsExcluded(FileSystemInfo fileSystemInfo, ExclusionFilters fi
...
+                        return targetInfo.LastWriteTimeUtc;
+                    }
+                }
+                catch (FileNotFoundException)
+                {
+                    // The file ceased to exist between LinkTarget and ResolveLinkTarget.
+                    // Try one more time, if it fails again just give up.
+                    if (!isSecondTry)
+                    {
+                        GetFileLinkTargetLastWriteTimeUtc(fileInfo, isSecondTry: true);
```
Comment (danmoseley) [inline src/libraries/Microsoft.Extensions.FileProviders.Physical/src/Internal/FileSystemInfoHelper.cs:68]: @Jozkee I'm missing something, shouldn't this return a value?

Comment (jeffhandley) [reply src/libraries/Microsoft.Extensions.FileProviders.Physical/src/Internal/FileSystemInfoHelper.cs:68]: 😳 I overlooked that too!

Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.FileProviders.Physical/src/Internal/FileSystemInfoHelper.cs:68]: @danmoseley great catch, I've sent https://github.com/dotnet/runtime/pull/57136 to address that

**Pull request merged**